### PR TITLE
add a new type StringArray

### LINKED ISSUE #89: Add a new type StringArray to handle multiple occurances of a flag without comma separation
Recently, there is a type `StringSlice` to handle the flag which is inputed multiple times. But it also handle each argument by comma separation without escaping. It is unexpected in `StringArray`.
For example
```
# input:
--value=one,two --value=three

# result by StringSlice
["one", "two", "three"]

# result expected by StringArray
["one,two", "three"]
```
A discussion about this function is [here](https://github.com/kubernetes/kubernetes/issues/27454#issuecomment-242714846)

### PR DESCRIPTION
**Pull request by wu8685**
fix #89

### CODE REVIEW

**Comment by wu8685** [comment]
It is not easy to get rid of CVS. I have to use CVS to combine all items of the array in `StringArray` into one string and split the string using CVS to get the []string value which will be returned by `GetStringArray()`.

Considering this [test case](https://github.com/wu8685/pflag/blob/52782456bfea413c905c62793d2561a25187e1c9/string_array_test.go#L156), I also have to use CVS to deal with `"`

File: string_array_test.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+		}
+	}
+}
+
+func TestSAWithSpecialChar(t *testing.T) {
+	var sa []string
+	f := setUpSAFlagSet(&sa)
+
+	in := []string{"one,two", `"three"`, `"four,five",six`, `"seven eight"`}
+	expected := []string{"one,two", `three`, `"four,five",six`, "seven eight"}
```
Comment (eparis) [inline string_array_test.go]: Is removing the `"` around `three` the correct thing to do with your new 'exact' flag parser?

Comment (wu8685) [inline string_array_test.go]: I guess. The same as `--value="seven eight"`. I suppose the quote here is to escape.

**Comment by fabianofranz** [comment]
I'd expect the `StringArray` implementation to be much simpler, really raw, just use the values as they come. I may be missing something, but I don't get why you still need CSV. For example, `Set` should be not more than taking the value and `append` to the slice

```
func (s *stringArrayValue) Set(val string) error {
  if !s.changed {
    *s.value = []string{val}
    s.changed = true
  } else {
    *s.value = append(*s.value, val)
  }
  return nil
}
```

**Comment by eparis** [comment]
@fabianofranz the problem is that the `GetStringArray()` function requires one to be able to parse the results of `String()`  which means the output of `String()` must be encoded in some sane way (CSV). If we don't encode on `Set` one would get a slightly different result via the variable bound to the flag and the result of `GetStringArray()`....

**Comment by wu8685** [comment]
@fabianofranz  If parsing argument as raw data, these input `--value=one,two --value="three" --value="seven eight"` would be parsed as `one,two`, `"three"`, `"seven eight"`. It is a little strange. But I am ok, if this is expected.

**Comment by fabianofranz** [comment]
@wu8685 That will only happen in tests. In the real world when using bash, if you take the raw values (I didn't test, but afaik that's what will happen)

`--value=one,two --value="three" --value="seven eight"` 

will become 

`one,two`, `three`, and `seven eight` 

Because bash already does some handling of quotes, double-quotes, spaces, etc; *before* sending the value to the program. If you do `echo "foo bar"` you get `foo bar`, not `"foo bar"`.

See, the reason for me arguing about this is that I wouldn't expect any additional magic from the non-CSV version. The program takes the values exactly as provided, I know my bash will do some pre-processing, but other than that I don't expect other surprises. Exactly like every other flag works -  the simple String one, for example, doesn't do any CSV, quotes or spaces processing. It just takes values as is. And this one, StringArray, is just a collection of raw Strings.

**Comment by wu8685** [comment]
@fabianofranz  Thanks for your explanation. I tested in bash. It proves you are right. I removed the CSV code within `Set()`

File: string_array.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+func (s *stringArrayValue) Type() string {
+	return "stringArray"
+}
+
+func (s *stringArrayValue) String() string {
+	str, _ := writeAsCSV(*s.value)
+	return "[" + str + "]"
+}
+
+func writeAsCSV(vals []string) (string, error) {
```
Comment (fabianofranz) [inline string_array.go]: Move this to `string_slice.go` and make `stringSliceValue.String()` to use this. Both `writeAsCSV` and `readAsCSV` will be in the same place.

**Comment by fabianofranz** [comment]
@wu8685 one minor comment then LGTM. @eparis PTAL?

**Pull request merged**